README.md
# voca-builder

voca-builder is an Emacs package that aimed to help you build up your vocabulary by automating the most step in the process, so that you actually spent time in learning new words. 

If you are learning a foreign language, you know how tedious and time-consuming it is to check a new word that you are not familiar with, it breaks you reading-flow, and therefore decrease your efficient. 

voca-builder provides excellent functionality that will do the checking for you, and shows the meaning as a popup above the text that you are reading. It also records the meaning and the sentence containing the word. All you need to do is press a key. 

It can also groups the vocabulary an later on, you can export new vocabularies based on tags or time period.  

For full documentation on this package, please see the wiki page.

# Demo

    (require 'voca-builder)
    (setq voca-builder/voca-file "~/git/Learning/Emacs_Voca/voca_example.org") 
    (setq voca-builder/current-tag "Demo")
    (global-set-key (kbd "<f4>") 'voca-builder/search-popup)
    
    (setq voca-builder/export-file "~/voca-builder-temp.org") 
    (voca-builder/extract-by-tags "Demo") 
    (voca-builder/extract-period "2015-01-05" "2015-04-01")

I moved the cursor to *c* under *search*, and press F4, it shows the short meaning, which has been added to a file, as the echo area shows. 
![img](./img/Poup-show-.png)
Export all the vocabularies that are tagged by *Demo*
![img](./img/Export-Tag-Demo.png)
Export all the vocabularies that are recorded between "2015-01-05" and "2014-04-01"
![img](./img/Export-Tag-Demo-Period-Range.png)

Here is a real example when I was reading an article.
![img](./img/Popup-real-example.png)

# It Make Your Life Easier

Image the process you want to look in a new word, it probably would like this: 

1.  put down the book you reading,
2.  find the dictionary or open the app,
3.  repeat step 1-2 few times if the vocabulary is quit complex,
4.  now you get the meaning, trying to comprehensive it,
5.  then you write it down in a note book, with the meaning,
6.  if you are very serious, you probably write down the sentence as an example,
7.  after all these, you go back to continue reading but lost the flow.

What if you can do the step 1-6 automatically, and immediately as you
thought you wanna see what's the meaning of a word? This is what
voca-bulder can do for you, all you need to do is press F4 (the
default setting), without leaving Emacs, searching and recording.
Sounds pretty cool is it?

Even more important, you can quary your vocbaulary database, and
search, for example, if you are prepare an interview in banking
sector, you can bring up all the banking related terminology, and
study them all together, or you wanna strength your memmoery and wants
to remmeber again before you forget, you can bring up all the words
you learnt last week, or this month etc.

# Human-Friendly Dictionary Database

take conciousness for example, definition from oxford is 

    [MASS NOUN] The state of being aware of and responsive to one’s surroundings:

while for vocabuary.com, it has more human friendly defition, which are for human, and edited by humans. 

    When you are awake and aware of your surroundings, that's consciousness. There are different types of consciousness, including social consciousness, being aware of injustices in society.

    The early Latin word conscius meant "with knowing," but it meant a shared knowledge, a meaning that was retained through the 16th Century. John Locke was the first to describe consciousness in relation to the individual, referring to it as “the perception of what passes in a man’s own mind." If you lose consciousness during the performance, we'll wake you up with some smelling salts!

# TODO TODO

1.  Currently voca-builder only supports English dictionary (www.vocabulary.com) and will add other languages later.
2.  Write Documentation


ChangeLog.org
* voca-builder 0.1.0 
[2015-06-08 Mon 19:49]
** Major changes
[2015-06-08 Mon 19:54]

For a word that has no meaning in the database,
=voca-builder/fetch-meaning= returns ='("No meaning found", "No
meaning found")= and =voca-builder/record-voca= won't update the
vocabulary file. 

** Minor changes 
[2015-06-08 Mon 19:56]

1. Namespace: =org-write-subtree= and =org-get-ts-for-subtree= become 
=voca-builder/org-write-subtree= and =voca-builder/org-get-ts-for-subtree=
2. Style: Place all trailing parentheses on a single line instead of distinct
   lines,  https://github.com/bbatsov/emacs-lisp-style-guide
3. change version number to 0.1.0


emacs_voca.el
;;; voca-builder.el --- Helps you build up your vocabulary
;;
;; Copyright (C) 2011-2014 Steve Purcell
;;
;; Author: Yi Tang <yi.tang.uk@me.com>
;; Keywords: English vocabulary 
;; Package-Requires: ((cl-lib "0.2"))
;; X-URL: http://github.com/yi-tang/voca-builder
;; URL: http://github.com/yi-tang/voca-builder
;; Version: DEV
;;
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; voca-builder is an Emacs package that aimed to help you build up your vocabulary by automating the most step in the process, so that you actually spent time in learning new words.
;; It will do the checking, and shows the meaning as a popup above the text. It also records the meaning and the sentence containing the word.  Finally, it can export vocabularies that have the same tags, or are between two date. 
;;; Use:
;;
;; To use voca-builder 
;;   (setq voca-builder/voca-file "~/.vocabulary.org")
;;   (setq voca-builder/export-file "~/.voca-builder-temp.org")
;;   (global-set-key (kbd "<f4>") 'voca-builder/search-popup) 
;;
;; To export all the vocabulary tagged by TLOTR 
;;   (voca-builder/extract-by-tags "TLOTR") , get all the vocabularies of The Lord of Rings.
;; To export all the vocabulary recored between 2015-01-01 and 2015-03-01
;;   (voca-builder/extract-period "2015-01-01" "2015-03-01")
;;
;;; Code:

;; requires

(require 'org)
(require 'popup)

(defgroup voca-builder nil
  "Group voca-builder entries according to their version control status."
  :prefix "voca-builder-"
  :group 'convenience)

(defcustom voca-builder/record-voca nil
  "If non-nil, record the vocabulary looked-into and save to voca-builder/voca-file."
  :type 'boolean
  :group 'voca-builder)

(defcustom voca-builder/voca-file "~/.vocabulary.org"
  "the vobulary file"
  :type 'file
  :group 'voca-builder)


(defcustom voca-builder/record-with-ts t
  "if non-nil, record the vocabulary with a timestamp."
  :type 'boolean
  :group 'voca-builder)

;; (defcustom voca-builder/ts-format nil
;;   "the timestsamp format for the records"
;;   :type 'string
;;   :group 'voca-builder)
(setq voca-builder/ts-format "[%Y-%m-%d %a %H:%M]") 

(defun voca-builder/record-vocabulary-p ()
  (if voca-builder/record-voca
      t
    f))

(defcustom voca-builder/popup-record-sentence nil
  "If non-nil, record the sentence which contain the word that was looked up."
  :type 'boolean
  :group 'voca-builder)
(setq voca-builder/popup-record-sentence t)

(defcustom voca-builder/popup-show-short-meaning nil
  "if non-nil, shows the short expnation of the vocabulary"
  :type 'boolean
  :group 'voca-builder)
(setq voca-builder/popup-short t)

(defcustom voca-builder/current-tag nil
  "if non-nil, add tags to the vocabulary node in org-mode"
  :type 'string
  :group 'voca-builder)
(setq voca-builder/current-tag "Gene") ;; 



;;;; config 
(setq voca-builder/popup-line-width 40)

;;;; section: html helper function 
(defun voca-builder/html-find-content-of-tags-by-points (s-tag e-tag)
  (interactive)
  (search-forward s-tag)
  (setq p1 (point))
  (search-forward e-tag)
  (backward-char (length e-tag))
  (setq p2 (point))
  (cons p1 p2)
  )

(defun voca-builder/make-url (voca)
  (concat "http://www.vocabulary.com/dictionary/" voca))

(defun voca-builder/fetch-meaning (voca)
  (setq url (voca-builder/make-url voca))
  (setq meaning (with-current-buffer
		    (url-retrieve-synchronously url)
		  (setq pts (voca-builder/html-find-content-of-tags-by-points "<p class=\"short\">" "</p>"))
		  (setq short-exp (buffer-substring (car pts) (cdr pts)))
		  (setq pts (voca-builder/html-find-content-of-tags-by-points "<p class=\"long\">" "</p>"))
		  (setq long-exp (buffer-substring (car pts) (cdr pts)))
		  (cons (decode-coding-string short-exp 'utf-8)
			(decode-coding-string long-exp 'utf-8))
		  
		  ))
  meaning 
  )

(defun voca-builder/voca-org-entry (voca exp extra)
  (setq ts (if voca-builder/record-with-ts
	       (format-time-string voca-builder/ts-format)))
  (setq tags (if (not (eq nil voca-builder/current-tag))
		 (concat ":" voca-builder/current-tag ":")))
  (concat "\n* " voca " " tags "\n" ts "\n\n" exp "\n\n" extra))

(defun voca-builder/record-voca (voca meaning extra)
  "reocrd the meaning of a vocabulary"
  (cond ((voca-builder/record-vocabulary-p)
	 (setq a-meaning (concat (car meaning) "\n\n" (cdr meaning)))
  	 (setq org-entry (voca-builder/voca-org-entry voca a-meaning extra))
  	 (message "%s" org-entry)
  	 (append-to-file org-entry nil voca-builder/voca-file)))
  )

(defun voca-builder/popup-clean-text (para)
  "remove html tags beofre display in emacs"
  (with-temp-buffer
    (insert para)
    (goto-char (point-min))
    (replace-string "<i>" "")
    (goto-char (point-min))
    (replace-string "</i>" "")
    (buffer-string))
  )

(defun voca-builder/search-popup ()
  (interactive)
  (setq voca (thing-at-point 'word))
  (if voca-builder/popup-record-sentence
      (setq sentence (thing-at-point 'sentence)))
  (setq meaning (voca-builder/fetch-meaning voca))
  (voca-builder/record-voca voca meaning sentence)
  (if voca-builder/popup-short
      (setq meaning (car meaning)))
  ;;  (message "%s" meaning)
  (setq b-meaning (voca-builder/popup-clean-text meaning))
  (popup-tip b-meaning :width voca-builder/popup-line-width)
  )



;;;; section: export 
(defun org-write-subtree ()
  "append current subtree to the voca-builder/export-file"
  (org-copy-subtree)
  (setq str (with-temp-buffer
	      (org-paste-subtree)
	      (buffer-string)))
  (append-to-file str nil voca-builder/export-file))

(defun voca-builder/extract-by-tags (tags)
  (setq res (org-map-entries 'org-write-subtree tags (list voca-builder/voca-file)))
  )


(defun org-get-ts-for-subtree ()
  "search timesamp in the current subtree, for example [2015-03-28 Sat 12:01], and parse it to date"
  (search-forward-regexp "[0-9]+-[0-9]+-[0-9]+")
  (beginning-of-line)
  (forward-char)
  (voca-builder/encode-date (buffer-substring (point) (+ (point) 10)))
  )

(defun voca-builder/encode-date (date1)
  "encode date
date: YYYY-MM-DD, for exmaple, 2015-12-01"
  (setq date1-s (split-string date1 "-"))
  (encode-time 0 0 0
	       (string-to-number (nth 2 date1-s))
	       (string-to-number (nth 1 date1-s))
	       (string-to-number (nth 0 date1-s))))

(defun voca-builder/extract-by-periods-helper ()
  (setq ts-sub-tree (org-get-ts-for-subtree))
  (setq p1 (time-less-p ts-sub-tree time2-internal))
  (setq p2 (time-less-p time1-internal ts-sub-tree))
  (if (and p1 p2)
      (org-write-subtree)))

(defun voca-builder/extract-period (p1 p2)
  "extract all vocabulary entries that are recorered between period p1 and p2.
period: YYYY-MM-DD, for exmaple, 2015-12-01"
  (interactive)
  (setq time1-internal (voca-builder/encode-date p1))
  (setq time2-internal (voca-builder/encode-date p2))
  (org-map-entries 'voca-builder/extract-by-periods-helper nil (list voca-builder/voca-file))
  )

(provide 'voca-builder)
;;;; voca-builder.el ends her


;;;; section: test
;; (setq voca-builder/voca-file "~/git/Learning/Emacs_Voca/voca_example.org") 
;; (setq voca-builder/export-file "~/voca-builder-temp.org") 
;; (voca-builder/extract-by-tags "Gene");; test for example, get all the vocabularies of The Lord of Rings.
;; (global-set-key (kbd "<f4>") 'voca-builder/search-popup)


voca_builder.el
;;; voca-builder.el --- Helps you build up your vocabulary
;;
;; Copyright (C) 2011-2014 Steve Purcell
;;
;; Author: Yi Tang <yi.tang.uk@me.com>
;; Keywords: English vocabulary 
;; Created: 28th March 2015
;; Package-Requires: (org popup)
;; X-URL: http://github.com/yitang/voca-builder
;; URL: http://github.com/yitang/voca-builder
;; Version: 0.0.20150328
;;
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; voca-builder is an Emacs package that aimed to help you build up your vocabulary by automating the most step in the process, so that you actually spent time in learning new words.
;; It will do the checking, and shows the meaning as a popup above the text. It also records the meaning and the sentence containing the word.  Finally, it can export vocabularies that have the same tags, or are between two date. 
;;; Use:
;;
;; To use voca-builder 
;;   (setq voca-builder/voca-file "~/.vocabulary.org")
;;   (setq voca-builder/export-file "~/.voca-builder-temp.org")
;;   (global-set-key (kbd "<f4>") 'voca-builder/search-popup) 
;;
;; To export all the vocabulary tagged by TLOTR 
;;   (voca-builder/extract-by-tags "TLOTR") , get all the vocabularies of The Lord of Rings.
;; To export all the vocabulary recored between 2015-01-01 and 2015-03-01
;;   (voca-builder/extract-period "2015-01-01" "2015-03-01")
;;
;;; Code:

;; requires

(require 'org)
(require 'popup)

(defgroup voca-builder nil
  "Group voca-builder entries according to their version control status."
  :prefix "voca-builder-"
  :group 'convenience)

(defcustom voca-builder/record-voca nil
  "If non-nil, record the vocabulary looked-into and save to voca-builder/voca-file."
  :type 'boolean
  :group 'voca-builder)

(defcustom voca-builder/voca-file "~/.vocabulary.org"
  "the vobulary file"
  :type 'file
  :group 'voca-builder)


(defcustom voca-builder/record-with-ts t
  "if non-nil, record the vocabulary with a timestamp."
  :type 'boolean
  :group 'voca-builder)

;; (defcustom voca-builder/ts-format nil
;;   "the timestsamp format for the records"
;;   :type 'string
;;   :group 'voca-builder)
(setq voca-builder/ts-format "[%Y-%m-%d %a %H:%M]") 

(defun voca-builder/record-vocabulary-p ()
  (if voca-builder/record-voca
      t
    f))

(defcustom voca-builder/popup-record-sentence nil
  "If non-nil, record the sentence which contain the word that was looked up."
  :type 'boolean
  :group 'voca-builder)
(setq voca-builder/popup-record-sentence t)

(defcustom voca-builder/popup-show-short-meaning nil
  "if non-nil, shows the short expnation of the vocabulary"
  :type 'boolean
  :group 'voca-builder)
(setq voca-builder/popup-short t)

(defcustom voca-builder/current-tag nil
  "if non-nil, add tags to the vocabulary node in org-mode"
  :type 'string
  :group 'voca-builder)
(setq voca-builder/current-tag "Gene") ;; 



;;;; config 
(setq voca-builder/popup-line-width 40)

;;;; section: html helper function 
(defun voca-builder/html-find-content-of-tags-by-points (s-tag e-tag)
  (interactive)
  (search-forward s-tag)
  (setq p1 (point))
  (search-forward e-tag)
  (backward-char (length e-tag))
  (setq p2 (point))
  (cons p1 p2)
  )

(defun voca-builder/make-url (voca)
  (concat "http://www.vocabulary.com/dictionary/" voca))

(defun voca-builder/fetch-meaning (voca)
  (setq url (voca-builder/make-url voca))
  (setq meaning (with-current-buffer
		    (url-retrieve-synchronously url)
		  (setq pts (voca-builder/html-find-content-of-tags-by-points "<p class=\"short\">" "</p>"))
		  (setq short-exp (buffer-substring (car pts) (cdr pts)))
		  (setq pts (voca-builder/html-find-content-of-tags-by-points "<p class=\"long\">" "</p>"))
		  (setq long-exp (buffer-substring (car pts) (cdr pts)))
		  (cons (decode-coding-string short-exp 'utf-8)
			(decode-coding-string long-exp 'utf-8))
		  
		  ))
  meaning 
  )

(defun voca-builder/voca-org-entry (voca exp extra)
  (setq ts (if voca-builder/record-with-ts
	       (format-time-string voca-builder/ts-format)))
  (setq tags (if (not (eq nil voca-builder/current-tag))
		 (concat ":" voca-builder/current-tag ":")))
  (concat "\n* " voca " " tags "\n" ts "\n\n" exp "\n\n" extra))

(defun voca-builder/record-voca (voca meaning extra)
  "reocrd the meaning of a vocabulary"
  (cond ((voca-builder/record-vocabulary-p)
	 (setq a-meaning (concat (car meaning) "\n\n" (cdr meaning)))
  	 (setq org-entry (voca-builder/voca-org-entry voca a-meaning extra))
  	 (message "%s" org-entry)
  	 (append-to-file org-entry nil voca-builder/voca-file)))
  )

(defun voca-builder/popup-clean-text (para)
  "remove html tags beofre display in emacs"
  (with-temp-buffer
    (insert para)
    (goto-char (point-min))
    (replace-string "<i>" "")
    (goto-char (point-min))
    (replace-string "</i>" "")
    (buffer-string))
  )

(defun voca-builder/search-popup ()
  (interactive)
  (setq voca (thing-at-point 'word))
  (if voca-builder/popup-record-sentence
      (setq sentence (thing-at-point 'sentence)))
  (setq meaning (voca-builder/fetch-meaning voca))
  (voca-builder/record-voca voca meaning sentence)
  (if voca-builder/popup-short
      (setq meaning (car meaning)))
  ;;  (message "%s" meaning)
  (setq b-meaning (voca-builder/popup-clean-text meaning))
  (popup-tip b-meaning :width voca-builder/popup-line-width)
  )



;;;; section: export 
(defun org-write-subtree ()
  "append current subtree to the voca-builder/export-file"
  (org-copy-subtree)
  (setq str (with-temp-buffer
	      (org-paste-subtree)
	      (buffer-string)))
  (append-to-file str nil voca-builder/export-file))

(defun voca-builder/extract-by-tags (tags)
  (setq res (org-map-entries 'org-write-subtree tags (list voca-builder/voca-file)))
  )


(defun org-get-ts-for-subtree ()
  "search timesamp in the current subtree, for example [2015-03-28 Sat 12:01], and parse it to date"
  (search-forward-regexp "[0-9]+-[0-9]+-[0-9]+")
  (beginning-of-line)
  (forward-char)
  (voca-builder/encode-date (buffer-substring (point) (+ (point) 10)))
  )

(defun voca-builder/encode-date (date1)
  "encode date
date: YYYY-MM-DD, for exmaple, 2015-12-01"
  (setq date1-s (split-string date1 "-"))
  (encode-time 0 0 0
	       (string-to-number (nth 2 date1-s))
	       (string-to-number (nth 1 date1-s))
	       (string-to-number (nth 0 date1-s))))

(defun voca-builder/extract-by-periods-helper ()
  (setq ts-sub-tree (org-get-ts-for-subtree))
  (setq p1 (time-less-p ts-sub-tree time2-internal))
  (setq p2 (time-less-p time1-internal ts-sub-tree))
  (if (and p1 p2)
      (org-write-subtree)))

(defun voca-builder/extract-period (p1 p2)
  "extract all vocabulary entries that are recorered between period p1 and p2.
period: YYYY-MM-DD, for exmaple, 2015-12-01"
  (interactive)
  (setq time1-internal (voca-builder/encode-date p1))
  (setq time2-internal (voca-builder/encode-date p2))
  (org-map-entries 'voca-builder/extract-by-periods-helper nil (list voca-builder/voca-file))
  )

(provide 'voca-builder)
;;;; voca-builder.el ends her


;;;; section: test
;; (setq voca-builder/voca-file "~/git/Learning/Emacs_Voca/voca_example.org") 
;; (setq voca-builder/export-file "~/voca-builder-temp.org") 
;; (voca-builder/extract-by-tags "Gene");; test for example, get all the vocabularies of The Lord of Rings.
;; (global-set-key (kbd "<f4>") 'voca-builder/search-popup)


voca-builder.el
;;; voca-builder.el --- Helps you build up your vocabulary
;;
;; Copyright (C) 2015 Yi Tang
;;
;; Author: Yi Tang <yi.tang.uk@me.com>
;; Keywords: English vocabulary 
;; Created: 28th March 2015
;; Package-Requires: ((popup "0.5.2"))
;; URL: https://github.com/yitang/voca-builder
;; Version: 0.1.0
;;
;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; voca-builder is an Emacs package that aimed to help you build up your
;; vocabulary by automating the most step in the process, so that you
;; actually spent time in learning new words.
;; It will do the checking, and shows the meaning as a popup above the
;; text. It also records the meaning and the sentence containing the word.
;; Finally, it can export vocabularies that have the same tags, or are
;; between two dates.
;;
;;; Use:
;;
;; To use voca-builder 
;;   (setq voca-builder/voca-file "~/.vocabulary.org")
;;   (setq voca-builder/export-file "~/.voca-builder-temp.org")
;;   (setq voca-builder/current-tag "Demo")
;;   (global-set-key (kbd "<f4>") 'voca-builder/search-popup) 
;;   
;; To export all the vocabulary tagged by TLOTR 
;;   (voca-builder/extract-by-tags "TLOTR") , get all the vocabularies tagged
;;   ;; by TLOTR,  The Lord of The Rings.
;; To export all the vocabulary recored between 2015-01-01 and 2015-03-01
;;   (voca-builder/extract-period "2015-01-01" "2015-03-01")
;;
;;; Code:

;; requires

(require 'org)
(require 'popup)

(defgroup voca-builder nil
  "Group voca-builder entries according to their version control status."
  :prefix "voca-builder-"
  :group 'convenience)

(defcustom voca-builder/voca-file "~/.vocabulary.org"
  "the file to store vocabularies" 
  :type 'file
  :group 'voca-builder)

(defcustom voca-builder/record-new-vocabulary t
  "If non-nil, record the new vocabulary that is checked and save the notes to voca-builder/voca-file."
  :type 'boolean
  :group 'voca-builder)

(defcustom voca-builder/record-with-ts t
  "if non-nil, record the vocabulary with a timestamp.

The timestamps are needed for export function"
  :type 'boolean
  :group 'voca-builder)

(defcustom voca-builder/popup-record-sentence t
  "If non-nil, record the sentence which contain the word that was looked into."
  :type 'boolean
  :group 'voca-builder)

(defcustom voca-builder/popup-show-short-meaning t
  "if non-nil, shows the short explnation of the vocabulary"
  :type 'boolean
  :group 'voca-builder)

(defcustom voca-builder/current-tag "Gene"
  "if non-nil, add tags to the vocabulary notes in org-mode"
  :type 'string
  :group 'voca-builder)

(defcustom voca-builder/popup-line-width 40
  "width of the popup menu"
  :type 'integer
  :group 'voca-builder)

;;;; section: functions 
(defun voca-builder/make-url (voca)
  (concat "http://www.vocabulary.com/dictionary/" voca))

(defun voca-builder/html-find-tag (tag &optional begining)
  "search for a html tag and return the point, if search filed, return the end of buffer point.
if begining is non-nil, return the point at the begining of the tag, instead of at the end"
  (cond ((search-forward tag nil t)
	 (if begining
	     (- (point) (length tag))
	   (point)))
	(t
	 ;;       (message "nothing found")
	 (point-max))))

(defun voca-builder/html-remove-emphasis-tags (a-string)
  "remove emphasis tags"
  (with-temp-buffer
    (insert a-string)
    (goto-char (point-min))
    (replace-string "<i>" "")
    (goto-char (point-min))
    (replace-string "</i>" "")
    (buffer-string)))

(defun voca-builder/html-find-content-of-tags (tag1 tag2)
  "It searchs the content that is wrapped by tag1 and tag2 in a HTML file, and return as a UTF-8 stirng. "
  (let* ((p-tag1 (voca-builder/html-find-tag tag1))
	 (p-tag2 (voca-builder/html-find-tag tag2 t))
	 (content (buffer-substring p-tag1 p-tag2))
	 (content-without-emphasis (voca-builder/html-remove-emphasis-tags content)))
    (decode-coding-string content-without-emphasis 'utf-8)))

(defun voca-builder/fetch-meaning (voca)
  "Parse the html content from www.vocabulary.com, and return the short and long meaning."
  (interactive)
  (with-current-buffer
      (url-retrieve-synchronously (voca-builder/make-url voca))
    (let* ((short-meaning (voca-builder/html-find-content-of-tags "<p class=\"short\">"
								  "</p>"))
	   (short-meaning (if (eq 0 (length short-meaning)) ;; if it has no short or long meaning. 
			      (voca-builder/html-find-content-of-tags "<meta name=\"description\" content =\""
								      "\" />")
			    short-meaning))	   
	   (long-meaning (voca-builder/html-find-content-of-tags "<p class=\"long\">"
								 "</p>")))
      (if (string-match-p "Try the world&#039;s fastest, smartest dictionary:" short-meaning)
	  (cons "No meaning found"
		"No meaning found")
	(cons short-meaning
	      long-meaning)))))

(defun voca-builder/voca-org-entry (voca exp extra)
  "create a org-mode sub-tree for the new vocabulary, with timetsamp and tags"
  (let ((ts (if voca-builder/record-with-ts
		(format-time-string "[%Y-%m-%d %a %H:%M]")))
	(tag (if voca-builder/current-tag
		 (concat ":" voca-builder/current-tag ":"))))
    (concat "\n* " voca " " tag "\n" ts "\n\n" exp "\n\n" extra)))

(defun voca-builder/record-voca (voca meaning extra)
  "save the vocabulary notes to file."
  (unless (string= "No meaning found" (car meaning))
    (cond (voca-builder/record-new-vocabulary
	   (let* ((string-meaning (concat
				   (car meaning)
				   "\n\n"
				   (cdr meaning)))
		  (org-entry (voca-builder/voca-org-entry voca string-meaning extra)))
	     (append-to-file org-entry nil voca-builder/voca-file))))))

(defun voca-builder/search-popup ()
  "search the word and shows the meaning in popup menu, may also save the notes"
  (interactive)
  (let* ((this-voca (thing-at-point 'word))
	 (this-sentence (if voca-builder/popup-record-sentence
			    (thing-at-point 'sentence)))
	 (meaning (voca-builder/fetch-meaning this-voca)))
    (if voca-builder/record-new-vocabulary
	(voca-builder/record-voca this-voca
				  meaning
				  this-sentence))
    (if voca-builder/popup-show-short-meaning
	(popup-tip (car meaning)
		   :width voca-builder/popup-line-width)
      (popup-tip (mapconcat 'identity
			    meaning
			    "\n")
		 :width voca-builder/popup-line-width))))

(defun voca-builder/search (this-voca)
  "search the word and shows the meaning in echo area, may also save the notes.
Back up function for voca-builder/search-popup."
  (interactive "sSearch for word: ")
  (let* ((this-sentence nil)
	 (meaning (voca-builder/fetch-meaning this-voca)))
    (if voca-builder/record-new-vocabulary
	(voca-builder/record-voca this-voca
				  meaning
				  this-sentence))
    (if voca-builder/popup-show-short-meaning
	(message "%s" (car meaning))
      (message "%s" (mapconcat 'identity
			       meaning
			       "\n")))))

;;;; section: export 
(defun voca-builder/org-write-subtree ()
  "append current subtree to the voca-builder/export-file"
  (org-copy-subtree)
  (let ((str (with-temp-buffer
	       (org-paste-subtree)
	       (buffer-string))))
    (append-to-file str nil voca-builder/export-file)))


(defun voca-builder/extract-by-tags (tags)
  "export all vocabulary records with tags"
  (interactive)
  (org-map-entries 'voca-builder/org-write-subtree tags (list voca-builder/voca-file)))


(defun voca-builder/org-get-ts-for-subtree ()
  "search timesamp in the current subtree, for example [2015-03-28 Sat 12:01], and parse it to date"
  (search-forward-regexp "[0-9]+-[0-9]+-[0-9]+")
  (beginning-of-line)
  (forward-char)
  (voca-builder/encode-date (buffer-substring (point) (+ (point) 10))))

(defun voca-builder/encode-date (date1)
  "encode date
date: YYYY-MM-DD, for exmaple, 2015-12-01"
  (let* ((date1-s (split-string date1 "-")))
    (encode-time 0 0 0
		 (string-to-number (nth 2 date1-s))
		 (string-to-number (nth 1 date1-s))
		 (string-to-number (nth 0 date1-s)))))

(defun voca-builder/extract-by-periods-helper ()
  "it is created for FUN in org-map-entries does not take arguments"
  (let* ((ts-sub-tree (voca-builder/org-get-ts-for-subtree))
	 (p1 (time-less-p ts-sub-tree time2-internal))
	 (p2 (time-less-p time1-internal ts-sub-tree)))
    (if (and p1 p2)
	(voca-builder/org-write-subtree))))

(defun voca-builder/extract-period (p1 p2)
  "extract all vocabulary entries that are recorered between period p1 and p2.
period: YYYY-MM-DD, for exmaple, 2015-12-01"
  (interactive)
  (let ((time1-internal (voca-builder/encode-date p1))
	(time2-internal (voca-builder/encode-date p2)))
    (org-map-entries 'voca-builder/extract-by-periods-helper nil (list voca-builder/voca-file))))

(provide 'voca-builder)

;;;; section: test 
;; (setq voca-builder/voca-file "~/vocabulary.org")
;; (setq voca-builder/current-tag "Demo")
;; (global-set-key (kbd "<f4>") 'voca-builder/search-popup)

;; (setq voca-builder/export-file "~/voca-builder-temp.org") 
;; (voca-builder/extract-by-tags "Demo") 
;; (voca-builder/extract-period "2015-01-05" "2015-04-01")

;;; voca-builder.el ends here
